refactor: Hook into CMake loading in a better spot. (#1989)
zmkfirmware/zmk: docs/docs/development/new-shield.md

## Overview

> This guide will walk through the steps necessary to add ZMK support for a keyboard the uses an addon MCU board (e.g. Pro Micro compatible) to provide the microprocessor.

Reviewer: I am aware some of these are unchanged from the previous version but might as well fix them while we are touching them?

```suggestion
This guide will walk through the steps necessary to add ZMK support for a keyboard that uses an add-on MCU board (e.g. Pro Micro compatible) to provide the microprocessor.
```

The high level steps are:

> - From a template, create a new [Zephyr module](https://docs.zephyrproject.org/3.2.0/develop/modules.html) to git repository contain one or more custom shields.

Reviewer: ```suggestion
- From a template, create a new [Zephyr module](https://docs.zephyrproject.org/3.2.0/develop/modules.html) housed in a git repository contain one or more custom shields.
```
Broken sentence, guessing your intention a bit.

- Create a new shield directory.
- Add the base Kconfig files.
- Add the shield overlay file to define the KSCAN driver for detecting key press/release.
- (Optional) Add the matrix transform for mapping KSCAN row/column values to sane key positions. This is needed for non-rectangular keyboards, or where the underlying row/column pin arrangement does not map one to one with logical locations on the keyboard.
- Add a default keymap, which users can override in their own configs as needed.
- Add a `<my_shield>.zmk.yml` metadata file to document the high level details of your shield, and the features it supports.
- Update the `build.yaml` file from the repository template to have some sample builds of the firmware to test.
- Add support for features such as encoders, OLED displays, or RGB underglow.

> It may be helpful to review the upstream [shields documentation](https://docs.zephyrproject.org/2.5.0/guides/porting/shields.html#shields) to get a proper understanding of the underlying system before continuing.

Reviewer: ```suggestion
It may be helpful to review the upstream [shields documentation](https://docs.zephyrproject.org/3.2.0/guides/porting/shields.html#shields) to get a proper understanding of the underlying system before continuing.
```

## Overview

This guide will walk through the steps necessary to add ZMK support for a keyboard that uses an add-on MCU board (e.g. Pro Micro compatible) to provide the microprocessor.

The high level steps are:

> - From a template, create a new [Zephyr module](https://docs.zephyrproject.org/3.2.0/develop/modules.html) housed in a git repository contain one or more custom shields.

Reviewer: ```suggestion
- From a template, create a new [Zephyr module](https://docs.zephyrproject.org/3.2.0/develop/modules.html) housed in a git repository containing one or more custom shields.
```

- Add a default keymap, which users can override in their own configs as needed.
- Add a `<my_shield>.zmk.yml` metadata file to document the high level details of your shield, and the features it supports.
- Update the `build.yaml` file from the repository template to have some sample builds of the firmware to test.
- Add support for features such as encoders, OLED displays, or RGB underglow.

It may be helpful to review the upstream [shields documentation](https://docs.zephyrproject.org/3.2.0/guides/porting/shields.html#shields) to get a proper understanding of the underlying system before continuing.

:::note
ZMK support for split keyboards requires a few more files than single boards to ensure proper connectivity between the central and peripheral units. Check the following guides thoroughly to ensure that all the files are in place.
:::

## New Zephyr Module Repository

The first step to creating the shield is to create a new Zephyr module repository from a template.

:::note

> This guide assume you already have a configured GitHub account. If you don't yet have one, go ahead and [sign up](https://github.com/join) before continuing.

Reviewer: ```suggestion
This guide assumes you already have a configured GitHub account. If you don't yet have one, go ahead and [sign up](https://github.com/join) before continuing.
```

:::

Follow these steps to create your new repository:

- Visit https://github.com/zmkfirmware/unified-zmk-config-template
- Click the green "Use this template" button
- In the drop down that opens, click "Use this template".
- In the following screen, provide the following information:
  - A repository name, e.g. `my-shield-module`.
  - A brief description, e.g. `ZMK Support For MyShield Keyboard`.
  - Select Public or Private, depending on your preference.

> - Click the green "Create repository button"

Reviewer: ```suggestion
- Click the green "Create repository" button
```

ZMK support for split keyboards requires a few more files than single boards to ensure proper connectivity between the central and peripheral units. Check the following guides thoroughly to ensure that all the files are in place.
:::

## New Zephyr Module Repository

The first step to creating the shield is to create a new Zephyr module repository from a template.

:::note
This guide assume you already have a configured GitHub account. If you don't yet have one, go ahead and [sign up](https://github.com/join) before continuing.
:::

Follow these steps to create your new repository:

- Visit https://github.com/zmkfirmware/unified-zmk-config-template
- Click the green "Use this template" button
- In the drop down that opens, click "Use this template".
- In the following screen, provide the following information:
  - A repository name, e.g. `my-shield-module`.
  - A brief description, e.g. `ZMK Support For MyShield Keyboard`.
  - Select Public or Private, depending on your preference.
- Click the green "Create repository button"

## New Shield Directory

:::note

> This guide describes how to add shield to an independently managed Zephyr module repository. This is the
> preferred way to handle boards and shields moving forward in ZMK, although the tooling to make this easier
> for users is still improving. ZMK does have a collection of boards/shields in the ZMK main repository, which are planned to be phased out, but until that is complete, there _may_ be a few select scenarios where adding your keyboard to ZMK itself is preferred. Due the volume of PRs and the focus on ZMK development, not merging of keyboard PRs, you are highly encouraged to use an out-of-tree Zephyr module repository to manage your definitions. Should you choose to try to get your keyboard included in ZMK main repository, the paths in the rest of the guide would be nested under the `app/` folder there instead. For example, `boards/shields/<keyboard_name>` should now be
> `app/boards/shields/<keyboard_name>`.

Reviewer: ```suggestion
This guide describes how to add a shield to an independently managed Zephyr module repository. This is the
preferred way to handle boards and shields moving forward in ZMK, although the tooling to make this easier
for users is still improving. ZMK does have a collection of boards/shields in the ZMK main repository, which are planned to be phased out, but until that is complete, there _may_ be a few select scenarios where adding your keyboard to ZMK itself is preferred. Due the volume of PRs and the focus of ZMK development not being merging of keyboard PRs, you are highly encouraged to use an out-of-tree Zephyr module repository to manage your definitions. Should you choose to try to get your keyboard included in ZMK main repository, the paths in the rest of the guide would be nested under the `app/` folder there instead. For example, `boards/shields/<keyboard_name>` should now be
`app/boards/shields/<keyboard_name>`.
```

The sentence in the middle seems broken, looks like we can remove "not merging of keyboard PRs, " alternative to the above.

## Default Keymap

Each keyboard should provide a default keymap to be used when building the firmware, which can be overridden and customized by user configs. For "shield keyboards", this should be placed in the `boards/shields/<shield_name>/<shield_name>.keymap` file. The keymap is configured as an additional devicetree overlay that includes the following:

> - A node with `compatible="zmk,keymap"` where each child node is a layer with a `bindings` array that binds each key position to a given behavior (e.g. key press, momentarily layer, etc).

Reviewer: ```suggestion
- A node with `compatible = "zmk,keymap"` where each child node is a layer with a `bindings` array that binds each key position to a given behavior (e.g. key press, momentary layer, etc).
```

You should place a properly named `foo.zmk.yml` file in the directory next to your other shield values, and fill it out completely and accurately. See [Hardware Metadata Files](/docs/development/hardware-metadata-files) for the full details.

## Build File

To help you test/verify your firmware, update the `build.yaml` to list your particular board/shield combinations you want built whenever changes are published to GitHub. Open `build.yaml` with your editor and add a combination, e.g.:

> ```

Reviewer: ```suggestion
```yaml
```

# This file generates the GitHub Actions matrix
# For simple board + shield combinations, add them
# to the top level board and shield arrays, for more
# control, add individual board + shield combinations to
# the `include` property, e.g:
#
# board: [ "nice_nano_v2" ]
# shield: [ "corne_left", "corne_right" ]
# include:
#   - board: bdn9_rev2
#   - board: nice_nano_v2
#     shield: reviung41
#
---
include:
  - board: nice_nano_v2
    shield: <my_shield>
```

> 

Reviewer: How about adding this paragraph here? I've seen this be a point of confusion a couple times. (couldn't put it in a suggestion with the code block.)

For split keyboards, you will need to specify the halves/siblings separately:
```yaml
[...]
include:
  - board: nice_nano_v2
    shield: <my_shield>_left
  - board: nice_nano_v2
    shield: <my_shield>_right
```

Author: Added.

## Testing

### GitHub Actions

Using GitHub Actions to build your new firmware can save you from doing any local [development setup](./setup.md),
at the expense of a longer feedback loop if there are issues. To push your changes and trigger a build:

- Add all your pending changes with `git add .`
- Commit your changes with `git commit -m "Initial shield"`
- Push the changes to GitHub with `git push`

Once pushed, click on the "Actions" tab of the repo you created in the first step, and you should see a new build running. If the build is successful, there will be a new `firmware.zip` artifact shown on the summary screen you can download that will contain the new `.uf2` files that can be flashed to the device.

### Local Build

:::note
To build locally, be sure you've followed the [development setup](./setup.md) guide first.
:::

Once you've fully created the new keyboard shield definition,
you should be able to test with a build command like:

```sh

> west build --pristine -b nice_nano_v2 -- -DSHIELD=my_board -DZMK_EXTRA_MODULES=/the/full/path/to/your/module

Reviewer: ```suggestion
west build --pristine -b nice_nano_v2 -- -DSHIELD=<my_shield> -DZMK_EXTRA_MODULES=/full/path/to/your/module
# replace <my_shield> with e.g. <my_shield>_left for split keyboards, then repeat for <my_shield>_right
```
I thought a comment like above might be a quick way to point out the difference without an additional paragraph.

```

> The above build command generates `build/zephyr/zmk.uf2` you can flash using the steps from the following section

Reviewer: ```suggestion
The above build command generates a `build/zephyr/zmk.uf2` file that you can flash using the steps from the following section. See the dedicated [building and flashing page](build-flash.md) for more details.
```
